lxd/instances_put.go

```diff
@@ -0,0 +1,319 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lxc/lxd/lxd/cgroup"
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+func coalesceErrors(errors map[string]error) error {
+	if len(errors) == 0 {
+		return nil
+	}
+
+	errorMsg := "The following instances failed to update state:\n"
+	for instName, err := range errors {
+		errorMsg += fmt.Sprintf(" - Instance: %s: %v\n", instName, err)
+	}
+
+	return fmt.Errorf("%s", errorMsg)
+}
+
+// Update all instance states
+func containersPut(d *Daemon, r *http.Request) response.Response {
+	project := projectParam(r)
+
+	c, err := instance.LoadByProject(d.State(), project)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	raw := api.InstancesPut{}
+	raw.Timeout = -1
+	if err := json.NewDecoder(r.Body).Decode(&raw); err != nil {
+		return response.BadRequest(err)
+	}
+
+	action := shared.InstanceAction(raw.Action)
+
+	var names []string
+	var instances []instance.Instance
+	for _, inst := range c {
+		switch action {
+		case shared.Freeze:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Restart:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Start:
+			if inst.IsRunning() {
+				continue
+			}
+		case shared.Stop:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Unfreeze:
+			if inst.IsRunning() {
+				continue
+			}
+		}
+
+		instances = append(instances, inst)
+		names = append(names, inst.Name())
+	}
+
+	// Don't mess with containers while in setup mode
+	<-d.readyChan
+
+	// Batch the actions.
+	var opType db.OperationType
+	var do func(*operations.Operation) error
+	switch action {
+	case shared.Start:
+		opType = db.OperationInstanceStart
+		do = func(op *operations.Operation) error {
+			failures := map[string]error{}
+			failuresLock := sync.Mutex{}
+			wgAction := sync.WaitGroup{}
+
+			for _, inst := range instances {
+				wgAction.Add(1)
+				go func(inst instance.Instance) {
+					defer wgAction.Done()
+
+					err = inst.Start(raw.Stateful)
+					if err != nil {
+						failuresLock.Lock()
+						failures[inst.Name()] = err
+						failuresLock.Unlock()
+					}
+				}(inst)
+			}
+
+			wgAction.Wait()
+			return coalesceErrors(failures)
+		}
+	case shared.Stop:
+		opType = db.OperationInstanceStop
+		if raw.Stateful {
+			do = func(op *operations.Operation) error {
+				failures := map[string]error{}
```

Comment:
@stgraber an observation here is that there's quite a lot of repetition on the `failures`, `failuresLock` and `wgAction` setup and usage. It may have been easier to initialise a buffered channel using a temporary struct with name and error result and then posted the result of each instance's action into it and then collect them at the end. But I note there do seem to be a few special cases further down that may have made this approach unfeasible.